### PR TITLE
Add ``histogram_matching`` functionality 

### PR DESCRIPTION
**Pull request by shijianjian**

### CODE REVIEW

**Comment by shijianjian** [comment]
Nope. It will not improve much on stitching results. Currently, it is not fully differentiable as the ```torch.unique``` part.

**Comment by edgarriba** [comment]
> Nope. It will not improve much on stitching results. Currently, it is not fully differentiable as the `torch.unique` part.

Aha, in any case I think this function is quite valuable

File: kornia/contrib/histogram_matching.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+def histogram_matching(source: torch.Tensor, template: torch.Tensor) -> torch.Tensor:
+    """Adjust the pixel values of an image to match its histogram towards a target image.
+
+    `Histogram matching <https://en.wikipedia.org/wiki/Histogram_matching>`_ is the transformation
+    of an image so that its histogram matches a specified histogram. In this implementation, the
+    histogram is computed over the flattened image array. Code referred to
+    `here <https://stackoverflow.com/questions/32655686/histogram-matching-of-two-images-in-python-2-x>`_.
+
+    Args:
+        source: Image to transform.
```
Comment (edgarriba) [inline kornia/contrib/histogram_matching.py:13]: Add expected image shapes

File: kornia/contrib/histogram_matching.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        template: Template image. It can have different dimensions to source.
+
+    Returns:
+        matched: The transformed output image.
```
Comment (edgarriba) [inline kornia/contrib/histogram_matching.py:17]:
```suggestion
        The transformed output image.
```
Add shape

File: kornia/contrib/histogram_matching.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    Note:
+        This function does not matches histograms element-wisely if input a batched tensor.
+    """
+
```
Comment (edgarriba) [inline kornia/contrib/histogram_matching.py:22]: Add some checkings

File: kornia/contrib/histogram_matching.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    Code referred to `here <https://github.com/pytorch/pytorch/issues/1552#issuecomment-926972915>`_.
+
+    Args:
+        x: the input tensor that needs to be interpolated.
+        xp: the x-coordinates of the referred data points.
+        fp: the y-coordinates of the referred data points, same length as ``xp``.
+
+    Returns:
+        The interpolated values, same shape as ``x``.
+    """
```
Comment (edgarriba) [inline kornia/contrib/histogram_matching.py:60]: Add some checkings

File: test/test_contrib.py [hunk 5]
```diff
@@ -368,3 +368,27 @@ def test_exception(self, device, dtype):
         stitcher = kornia.contrib.ImageStitcher(matcher).to(device=device, dtype=dtype)
         with pytest.raises(RuntimeError):
             stitcher(input1, input2)
+
+
+class TestHistMatch:
```
Comment (edgarriba) [inline test/test_contrib.py:373]: Add value tests

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: test/test_contrib.py [hunk 6]
```diff
@@ -368,3 +368,35 @@ def test_exception(self, device, dtype):
...
+        fp = torch.tensor([4, 2, 0], device=device, dtype=dtype)
+        x = torch.tensor([4, 5, 6], device=device, dtype=dtype)
+        x_hat_expected = torch.tensor([-2., -4., -6.], device=device, dtype=dtype)
+        x_hat = kornia.contrib.interp(x, xp, fp)
+        assert (x_hat_expected == x_hat).all()
+
+    def test_histmatch(self, device, dtype):
+        torch.manual_seed(44)
+        src = torch.randn(1, 4, 4, device=device, dtype=dtype)
+        dst = torch.randn(1, 16, 16, device=device, dtype=dtype)
```
Comment (shijianjian) [inline test/test_contrib.py:386]:
```suggestion
        # generate random value by CPU.
        src = torch.randn(1, 4, 4).to(device=device, dtype=dtype)
        dst = torch.randn(1, 16, 16).to(device=device, dtype=dtype)
```

Comment (shijianjian) [reply test/test_contrib.py:386]: @edgarriba Can you commit this? It somehow claims that I do not have write permission to commit the code changes. Any ideas?

File: test/test_contrib.py [hunk 7]
```diff
@@ -368,3 +368,35 @@ def test_exception(self, device, dtype):
...
+        out = kornia.contrib.histogram_matching(src, dst)
+        exp = torch.tensor([[
+            [0.9356, 0.8270, 1.3687, 0.5640],
+            [0.6273, 0.9119, 0.4965, 0.4020],
+            [0.4353, 0.1475, 0.3384, 0.2580],
+            [0.0606, 0.7531, 0.2139, 0.6932]
+        ]])
```
Comment (edgarriba) [inline test/test_contrib.py:393]:
```suggestion
        ]], device=device, dtype=dtype)
```

File: test/test_contrib.py (hunk 6 above)
Comment (edgarriba) [reply test/test_contrib.py:386]: What it differs from the other?

**Review by edgarriba — APPROVED** [review]
LGTM. Merge once green lights

**Comment by edgarriba** [comment]
@shijianjian can we have small example here ? @ducha-aiki how could this be used for template tracking ?

**Pull request merged**